File: src/test/java/de/schramm/royalbash/tdd/core/creature/GoblinRaiderTest.java
package de.schramm.royalbash.tdd.core.creature;

import de.schramm.royalbash.tdd.core.Context;
import de.schramm.royalbash.tdd.core.Game;
import de.schramm.royalbash.tdd.core.Player;
import de.schramm.royalbash.tdd.core.Spot;
import de.schramm.royalbash.tdd.core.card.NoOpCard;
import lombok.val;
import org.junit.Test;

import static org.assertj.core.api.Assertions.assertThat;

public class GoblinRaiderTest {

    @Test
    public void should_be_placeable_on_spot() {

        // Given
        val testee = GoblinRaider.builder().build();

        // When
        val isPlaceableOnSpot = testee.canBePlacedOnSpot();

        // Then
        assertThat(isPlaceableOnSpot).isTrue();
    }

    @Test
    public void should_deliver_hitpoints() {

        // Given
        val testee = GoblinRaider.builder()
                .hitpoints(2)
                .build();

        // When
        val hitpoints = testee.getHitpoints();

        // Then
        assertThat(hitpoints).isEqualTo(2);
    }

    @Test
    public void should_deliver_attack() {

        // Given
        val testee = GoblinRaider.builder()
                .attack(1)
                .build();

        // When
        val attack = testee.getAttack();

        // Then
        assertThat(attack).isEqualTo(1);
    }

    @Test
    public void should_discard_a_card_of_traget_player_when_placed_on_spot() {

        // Given
        val testee = GoblinRaider.builder().build();
        val spot = Spot.builder().build();
        val player1 = Player.builder()
                .spot(spot)
                .build();
        val player2 = Player.builder()
                .handcard(NoOpCard.builder().build())
                .handcard(NoOpCard.builder().build())
                .build();
        val game = Game.builder()
                .player1(player1)
                .player2(player2)
                .build();
        val context = Context.builder()
                .game(game)
                .owner(player1)
                .targetPlayer(player2)
                .targetSpot(spot)
                .build();

        // When
        val updatedGame = testee.invoke(context);

        // Then
        assertThat(updatedGame.getPlayer2().getHandcards()).hasSize(1);
    }

}

File: backend/src/test/kotlin/de/schramm/royalbash/domain/card/creature/GoblinRaiderTest.kt
package de.schramm.royalbash.domain.card.creature

import de.schramm.royalbash.domain.Context
import de.schramm.royalbash.domain.Game
import de.schramm.royalbash.domain.Player
import de.schramm.royalbash.domain.Spot
import de.schramm.royalbash.domain.card.CardMock
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test


class GoblinRaiderTest {

    @Test
    fun should_discard_a_card_of_traget_player_when_placed_on_spot() {

        // Given
        val testee = GoblinRaider(
                "Id 1",
                hitpoints = 1,
                attack = 1,
                cost = 1)
        val spot = Spot(id = "spot")
        val player1 = Player("Id 2", spots = listOf(spot))
        val player2 = Player("Id 3", handcards = listOf(CardMock("Id 4"), CardMock("Id 5")))
        val game = Game("Id 6", player1 = player1, player2 = player2)
        val context = Context(
                game,
                player1,
                targetPlayer = player2,
                targetSpot = spot)

        // When
        val updatedGame = testee.invoke(context)

        // Then
        assertThat(updatedGame.player2.handcards).hasSize(1)
    }

}